kedro/extras/datasets/email/message_dataset.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2020 QuantumBlack Visual Analytics Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND
+# NONINFRINGEMENT. IN NO EVENT WILL THE LICENSOR OR OTHER CONTRIBUTORS
+# BE LIABLE FOR ANY CLAIM, DAMAGES, OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF, OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# The QuantumBlack Visual Analytics Limited ("QuantumBlack") name and logo
+# (either separately or in combination, "QuantumBlack Trademarks") are
+# trademarks of QuantumBlack. The License does not grant you any right or
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+# or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""``MessageDataSet`` loads/saves an email message from/to a file
+using an underlying filesystem (e.g.: local, S3, GCS). It uses the
+``email`` package in the standard library to manage email messages.
+"""
+from copy import deepcopy
+from email.generator import Generator
+from email.message import Message
+from email.parser import Parser
+from email.policy import default
+from pathlib import PurePosixPath
+from typing import Any, Dict
+
+import fsspec
+
+from kedro.io.core import (
+    AbstractVersionedDataSet,
+    DataSetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+
+class MessageDataSet(
```

Comment:
Maybe we could name this `EmailMessageDataSet`? In the future, there might be `SlackMessageDataSet`, etc.

Reply:
I'm open to it. My logic was that this is already namespaced `email` (i.e. they would be distinguished in the catalog as `email.MessageDataSet` vs. `slack.MessageDataSet`). No strong feelings from my end either way.